src/server/editorServices.ts

```diff
@@ -308,7 +308,7 @@ namespace ts.server {
     }
 
     interface AssignProjectResult extends OpenConfiguredProjectResult {
-        defaultConfigProject: ConfiguredProject | undefined;
+        retainProjects: readonly ConfiguredProject[] | ConfiguredProject | undefined;

```

Comment:
>retainProjects [](start = 8, length = 14)

Singleton and undefined are included because they're cheaper to store than the corresponding lists?  If so, we don't already have a helper type for that?

Reply:
Havent found it in core so far

Reply:
There are helpers that operate on `T | readonly T[]`, but I also can't find an attempt to formalize the pattern.